File: src/main/java/no/nav/syfo/service/BrukertilgangService.java
package no.nav.syfo.service;

import no.nav.syfo.aktorregister.AktorregisterConsumer;
import no.nav.syfo.aktorregister.domain.Fodselsnummer;
import no.nav.syfo.brukertilgang.BrukertilgangConsumer;
import no.nav.syfo.consumer.ws.PersonConsumer;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;

import javax.inject.Inject;
import javax.ws.rs.ForbiddenException;

import static no.nav.syfo.config.CacheConfig.CACHENAME_TILGANG_IDENT;

@Service
public class BrukertilgangService {

    private final AktorregisterConsumer aktorregisterConsumer;
    private final BrukertilgangConsumer brukertilgangConsumer;
    private final PersonConsumer personConsumer;

    @Inject
    public BrukertilgangService(
            AktorregisterConsumer aktorregisterConsumer,
            BrukertilgangConsumer brukertilgangConsumer,
            PersonConsumer personConsumer
    ) {
        this.aktorregisterConsumer = aktorregisterConsumer;
        this.brukertilgangConsumer = brukertilgangConsumer;
        this.personConsumer = personConsumer;
    }

    public boolean harTilgangTilOppslaattBruker(String innloggetIdent, String fnr) {
        String oppslaattAktoerId = aktorregisterConsumer.getAktorIdForFodselsnummer(new Fodselsnummer(fnr));
        try {
            return !(sporOmNoenAndreEnnSegSelvEllerEgneAnsatte(innloggetIdent, fnr)
                    || personConsumer.erBrukerKode6(oppslaattAktoerId));
        } catch (ForbiddenException e) {
            return false;
        }
    }

    @Cacheable(cacheNames = CACHENAME_TILGANG_IDENT, key = "#innloggetIdent.concat(#oppslaattFnr)", condition = "#innloggetIdent != null && #oppslaattFnr != null")
    public boolean sporOmNoenAndreEnnSegSelvEllerEgneAnsatte(String innloggetIdent, String oppslaattFnr) {
        return !(oppslaattFnr.equals(innloggetIdent) || brukertilgangConsumer.hasAccessToAnsatt(oppslaattFnr));
    }
}

File: src/main/kotlin/no/nav/syfo/brukertilgang/BrukertilgangService.kt
package no.nav.syfo.brukertilgang

import no.nav.syfo.aktorregister.domain.Fodselsnummer
import no.nav.syfo.config.CacheConfig
import no.nav.syfo.pdl.PdlConsumer
import no.nav.syfo.pdl.isKode6
import org.springframework.cache.annotation.Cacheable
import org.springframework.stereotype.Service
import javax.inject.Inject
import javax.ws.rs.ForbiddenException

@Service
class BrukertilgangService @Inject constructor(
        private val brukertilgangConsumer: BrukertilgangConsumer,
        private val pdlConsumer: PdlConsumer
) {
    fun harTilgangTilOppslaattBruker(innloggetIdent: String, fnr: String): Boolean {
        return try {
            !(sporOmNoenAndreEnnSegSelvEllerEgneAnsatte(innloggetIdent, fnr)
                    || pdlConsumer.person(Fodselsnummer(fnr))?.isKode6() == true)
        } catch (e: ForbiddenException) {
            false
        }
    }

    @Cacheable(cacheNames = [CacheConfig.CACHENAME_TILGANG_IDENT], key = "#innloggetIdent.concat(#oppslaattFnr)", condition = "#innloggetIdent != null && #oppslaattFnr != null")
    fun sporOmNoenAndreEnnSegSelvEllerEgneAnsatte(innloggetIdent: String, oppslaattFnr: String): Boolean {
        return !(oppslaattFnr == innloggetIdent || brukertilgangConsumer.hasAccessToAnsatt(oppslaattFnr))
    }

}
